test/Microsoft.AspNetCore.Razor.Tools.Test/RequestDispatcherTest.cs

```diff
@@ -0,0 +1,552 @@
+                    return EmptyBuildResponse;
+                };
+            });
+
+            var keepAlive = TimeSpan.FromSeconds(1);
+            var eventBus = new TestableEventBus();
+            var dispatcher = new DefaultRequestDispatcher(connectionHost, compilerHost, CancellationToken.None, eventBus, keepAlive);
+
+            // Act
+            dispatcher.Run();
+
+            // Assert
+            Assert.Equal(count, eventBus.CompletedCount);
+            Assert.True(eventBus.LastProcessedTime.HasValue);
+            Assert.True(eventBus.HitKeepAliveTimeout);
+        }
+
+        /// <summary>
+        /// Ensure server respects keep alive and shuts down after processing simultaneous connections.
+        /// </summary>
+        [Fact]
+        public async Task KeepAliveAfterSimultaneousConnections()
+        {
+            // Arrange
+            var totalCount = 2;
+            var readySource = new TaskCompletionSource<bool>();
+            var list = new List<TaskCompletionSource<bool>>();
+            var connectionHost = new Mock<ConnectionHost>();
+            connectionHost
+                .Setup(x => x.WaitForConnectionAsync(It.IsAny<CancellationToken>()))
+                .Returns((CancellationToken ct) =>
+                {
+                    if (list.Count < totalCount)
+                    {
+                        var source = new TaskCompletionSource<bool>();
+                        var connectionTask = CreateConnectionWithEmptyBuildRequest(c =>
+                        {
+                            c.WaitForDisconnectAsyncFunc = _ => source.Task;
+                        });
+                        list.Add(source);
+                        return connectionTask;
+                    }
+
+                    readySource.SetResult(true);
+                    return new TaskCompletionSource<Connection>().Task;
+                });
+
+            var compilerHost = CreateCompilerHost(c =>
+            {
+                c.ExecuteFunc = (req, ct) =>
+                {
+                    return EmptyBuildResponse;
+                };
+            });
+
+            var keepAlive = TimeSpan.FromSeconds(1);
+            var eventBus = new TestableEventBus();
+            var dispatcherTask = Task.Run(() =>
+            {
+                var dispatcher = new DefaultRequestDispatcher(connectionHost.Object, compilerHost, CancellationToken.None, eventBus, keepAlive);
+                dispatcher.Run();
+            });
+
+            await readySource.Task;
+            foreach (var source in list)
+            {
+                source.SetResult(true);
+            }
+
+            // Act
+            await dispatcherTask;
+
+            // Assert
+            Assert.Equal(totalCount, eventBus.CompletedCount);
+            Assert.True(eventBus.LastProcessedTime.HasValue);
+            Assert.True(eventBus.HitKeepAliveTimeout);
+        }
+
+        [Fact]
+        public void ClientExceptionShouldBeginShutdown()
```

Comment:
This seems like a pretty good start. I want to spend a little time thinking about the scenarios covered here to see if we cover everything.